padertorch/train/trainer.py

```diff
@@ -362,39 +469,65 @@ def optimizer_zero_grad(self):
             self.optimizer.zero_grad()
 
     def optimizer_step(self):
+        summary = self.clip_grad({})
+
+        # Add learning rate to the summary
+        if isinstance(self.optimizer, dict):
+            for key, optim in self.optimizer.items():
+                for i, param_group in enumerate(optim.optimizer.param_groups):
+                    summary['scalars'][f'lr/{key}/param_group_{i}'] = param_group['lr']
+        else:
+            for i, param_group in enumerate(self.optimizer.optimizer.param_groups):
+                summary['scalars'][f'lr/param_group_{i}'] = param_group['lr']
+
+        # Do the actual optimization
         if isinstance(self.optimizer, dict):
             for opti in self.optimizer.values():
                 opti.step()
         else:
             self.optimizer.step()
 
-    def train_step(self, example, optimize=True):
-
-        model_out, review = self.step(example, self.train_timer)
-
-        with self.train_timer['time_per_backward']:
-            self.backward(review)
-            if optimize:
-                review = self.clip_grad(review)
-                self.optimizer_step()
-                self.optimizer_zero_grad()
+        self.optimizer_zero_grad()
+        return summary
 
-        return model_out, review
+    def train_step(self, model, example, device):
+        return self.step(model, example, self.train_timer, device)
 
-    def validation_step(self, example):
-        return self.step(example, self.validate_timer)
+    def validation_step(self, model, example, device):
+        # [1:] -> ignore the loss. Is already in scalars.
+        return self.step(model, example, self.validate_timer, device)[1:]
 
-    def step(self, example, timer):
+    def step(self, model, example, timer, device):
         # TODO: Backup OutOfMemory
+        # with self.train_timer['time_per_step']:
```

Comment:
Time per step is no longer reported?

Reply:
I dropped it, because it is relative similar to `time_per_iteration` ( = `time_per_data_loading` + `time_per_step`). Since backward is no longer in the step, it cannot directly be measured.